python/ray/_private/runtime_env/working_dir.py

```diff
@@ -178,9 +182,78 @@ def modify_context(
                 "downloading or unpacking the working_dir."
             )
 
+        # Use placeholder here and will replace it by `pre_worker_startup`.
         if not _WIN32:
-            context.command_prefix += ["cd", str(local_dir), "&&"]
+            context.command_prefix += [
+                "cd",
+                WorkingDirPlugin.working_dir_placeholder,
+                "&&",
+            ]
         else:
             # Include '/d' incase temp folder is on different drive than Ray install.
-            context.command_prefix += ["cd", "/d", f"{local_dir}", "&&"]
-        set_pythonpath_in_context(python_path=str(local_dir), context=context)
+            context.command_prefix += [
+                "cd",
+                "/d",
+                WorkingDirPlugin.working_dir_placeholder,
+                "&&",
+            ]
+        set_pythonpath_in_context(
+            python_path=WorkingDirPlugin.working_dir_placeholder, context=context
+        )
+        context.cwd = WorkingDirPlugin.working_dir_placeholder
+        context.symlink_dirs_to_cwd.append(str(local_dir))
+
+    async def pre_worker_startup(
+        self,
+        runtime_env: "RuntimeEnv",  # noqa: F821
+        context: RuntimeEnvContext,
+        worker_id: str,
+    ) -> None:
+        if not runtime_env.working_dir():
+            return
+        default_logger.info(f"Creating working dir for worker {worker_id}.")
+        working_dir = os.path.join(self._working_dirs, worker_id)
+        os.makedirs(working_dir, exist_ok=True)
+        context.cwd = working_dir
+        # Replace the placeholder with the real working dir.
+        for i, prefix in enumerate(context.command_prefix):
+            context.command_prefix[i] = prefix.replace(
+                WorkingDirPlugin.working_dir_placeholder, working_dir
+            )
+        for k, v in context.env_vars.copy().items():
+            context.env_vars[k] = v.replace(
+                WorkingDirPlugin.working_dir_placeholder, working_dir
+            )
+        if "container_command" in context.container:
+            for i, command_str in enumerate(context.container["container_command"]):
+                context.container["container_command"][i] = command_str.replace(
+                    WorkingDirPlugin.working_dir_placeholder, working_dir
+                )
+        # Add symbol links to the working dir.
+        for symlink_dir in context.symlink_dirs_to_cwd:
+            for name in os.listdir(symlink_dir):
+                src_path = os.path.join(symlink_dir, name)
+                link_path = os.path.join(working_dir, name)
```

Comment:
This means if Ray task writes to the symlink dir, we'll still have the same issue?

Reply:
Discussed offline: ideally symlink dir should be read-only but we keep it writable for backward compatibility.